### ISSUE TITLE
Question about calculating the inter-agent replusion

### ISSUE THREAD

**Issue by ljt1018** [realcrane/Human-Trajectory-Prediction-via-Neural-Social-Physics#14]
Hi, thanks for your great work and for releasing the code!
When calculating the replusive force on agent i from other adjacent agents, it requires future position from the neighboring agents. How  do you acquire the future position of neighboring agents? Are you directly utilizing the ground truth position value from the dataset for the nearby agents?

**Comment by sunny123123** [comment]
@ljt1018 Yes, he directly utilize the ground truth position. I'm skeptical of the results of this paper

**Comment by mulplue** [comment]
In test() function, the supplement input is `supplement[:, 7, :, :]`.
So I think it doesn't make use of future trajectories when computing replusive force.

**Comment by sunny123123** [comment]
@mulplue No，you're not very careful. 
<img width="1338" alt="image" src="https://user-images.githubusercontent.com/10163641/225513629-5843d5e2-ce6b-4164-a95f-203e31a6dd17.png">
please look at the second red box.

**Comment by mulplue** [comment]
> 
You're right, it's in the test function, the code leak the ground truth.
Besides, I found that the ADE and FDE error output using the saved_model is not same as the paper results:
In my test on SDD with test function of  `train_nsp_w.py`, I get `ADE=9.790649215889587 & FDE=11.647142716725744` 
but in the paper experiments, the result is `ADE=6.52 & FDE=10.61` 
Do you encounter the same problem?

**Comment by sunny123123** [comment]
@mulplue Maybe add cvae moudle can achieve the author's result. It  doesn't matter how good the results based on error data used.

**Comment by JiangbeiYue** [comment]
@ljt1018 @mulplue @sunny123123 Hi all, Thanks for your questions, and we are sorry for the late reply.

First, the repulsive force is not computed based on the future positions of the neighboring agents. It is computed based on the current positions of the neighboring agents.

Further, we didn’t utilize the ground truth positions during testing. When we calculate the repulsive force on the agent i from its neighbors, at each step, we first update the positions of all agents.

In our method, there are two types of neighboring agents for any agent i at any time t. The first type is the agents with complete 20-frame trajectories (past 8 frames and future 12 frames). Their positions are predicted on the fly at each step (see the code in the model_nsp_wo.py at lines 381-383). The second type has less than 20 frames in the data. While previous methods chose to drop them such as Social GAN, we regard them as observed moving obstacles as part of the environment (see the code in line 384 of the model_nsp_wo.py). Overall, for example, we have 10 agents in a scene, and the target agent is the 4th agent. Agents 1-6 have complete 20-frame trajectories and the remaining 4 agents don’t (e.g., they only appear in 2 or 3 frames around the boundary of the scene). During prediction, we predict the positions of Agents 1-6 at each step and regard the remaining 4 agents as moving obstacles whose positions are in the supplement. Note that they are not considered as ground truth because we do not predict their trajectories and do not compute ADE/FDE on them.

**Comment by mulplue** [comment]
@JiangbeiYue Thanks for your reply!
You mentioned that "at each step, we first update the positions of all agents" during test, but I can't find the related code, can you tell me the location of this computation? 
Thanks very much!

**Comment by centiLinda** [comment]
@JiangbeiYue Understand that supplement[:, 7+t+1, :, :] is only used for moving obstacles (agents with incomplete trajectories). These are not agents we're interested in predicting so it doesn't directly leak the ground truth.

However, using ground truth 'moving obstacles' positions still makes it unfair for other methods. The inputs for most methods are the ground truth history trajectories and the images at any step. On top of these, your method makes use of ground truth obstacle positions at any step, which I don't think should be assumed available. 

Perhaps a better choice is to use a pretrained detection network to identify pedestrians from the image, filter out those we're predicting, and the rest can be treated as moving obstacles.

I will test your model without ground truth moving obstacles and see if there's a performance drop.

**Comment by JiangbeiYue** [comment]
Hi,

Previous methods didn’t discuss the optimality of 20 frames, which suggests this might be an arbitrary choice or the result of parameter tuning. Therefore, discarding trajectories that are shorter than 20 frames is not necessarily the only correct way to use the dataset, or the only fair way for comparison.  

In addition, to test if a model has truly learned the correct steering behavior, a pedestrian needs to observe the surrounding pedestrians. Imagine there are two pedestrians A and B. They walk towards one another initially then steer away to avoid collisions. In this case, B disappears after steering out of the camera view, so its trajectory is shorter than 20 frames while A has 20 frames. During testing, the model tries to predict the A’s trajectory. In our setting regarding trajectories shorter than 20 frames as moving obstacles, NSP will predict steering when B is present, or not steering otherwise. This is the correct behavior, i.e. no steering when there are no imminent collisions. If B is discarded, this creates an inconsistent testing scenario: if the model predicts steering, it produces a smaller error but wrong behavior; otherwise a bigger error but correct behavior. 

Due to the above reasons, we designed our comparison setting to be as fair as possible.
